File: backend/src/Services/PollService.cs
using API.Entities;
using API.Interfaces;
using API.Types;

namespace API.Services;

public class PollService(IPollRepository pollRepository) {

    public async Task<Page<Poll>> GetAll(string filter, int page, int limit) {

        Page<Poll> polls = await pollRepository.GetAll(filter, page, limit);

        return polls;

    }

    public async Task<Page<Poll>> GetPollsByBuilding(int buildingId, string filter, int page, int limit) {

        Page<Poll> polls = await pollRepository.GetByBuildingId(buildingId, filter, page, limit);

        return polls;

    }

    public async Task<Poll?> GetPollById(int id) {

        Poll? poll = await pollRepository.GetById(id);

        return poll;

    }

    public async Task<Poll?> CreatePoll(int buildingId, string title, bool isActive) {

        Poll poll = new() {
            BuildingId = buildingId,
            Title = title,
            IsActive = isActive
        };

        try {
            return await pollRepository.Create(poll);
        } catch {
            return null;
        }

    }

    public async Task<Poll?> UpdatePoll(int id, int? buildingId, string? title, bool? isActive) {

        Poll? poll = await pollRepository.GetById(id);

        if(poll == null) {
            return null;
        }

        poll.BuildingId = buildingId ?? poll.BuildingId;
        poll.Title = title ?? poll.Title;
        poll.IsActive = isActive ?? poll.IsActive;

        try {
            return await pollRepository.Update(poll);
        } catch {
            return null;
        }

    }

    public async Task<Poll?> DeletePoll(int id) {

        Poll? poll = await pollRepository.GetById(id);

        if(poll == null) {
            return null;
        }

        try {
            return await pollRepository.Delete(poll);
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/poll.service.ts
import { APIError } from "@/models/api-error.models";
import { Page } from "@/models/page";
import { Poll } from "@/models/poll.model";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";

export class PollService extends Service {

    static async getPolls(filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Poll> | APIError>("/api/poll", {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async getPollsByBuilding(buildingId : number, filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Poll> | APIError>(`/api/poll/building/${buildingId}`, {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createPoll(buildingId : number, title : string) {

        const response = await this.axios.post<Poll | APIError>("/api/poll", {
            buildingId,
            title
        });

        return response.data;

    }

    static async updatePoll(id : number, buildingId : optional<number>, title : optional<string>, isActive : optional<boolean>) {

        const response = await this.axios.put<Poll | APIError>(`/api/poll/${id}`, {
            buildingId,
            title,
            isActive
        });

        return response.data;

    }

    static async deletePoll(id : number) {

        const response = await this.axios.delete<Poll | APIError>(`/api/poll/${id}`);

        return response.data;

    }

}